addons/binding/org.openhab.binding.konnected/src/main/java/org/openhab/binding/konnected/handler/KonnectedHandler.java

```diff
@@ -0,0 +1,330 @@
+
+    @Override
+    public synchronized void channelLinked(ChannelUID channel) {
+        // adds linked channels to list based on last value of Channel ID
+        // which is set to a number
+        logger.debug("Channel {} has been linked", channel.getId());
+        setisAct();
+        // get the zone number in integer form
+        Integer zone = Integer.parseInt(channel.getId().substring((channel.getId().length() - 1)));
+
+        // convert the zone to the pin based on value at index of zone
+        Integer pin = Arrays.asList(PIN_TO_ZONE).get(zone);
+
+        // if the pin is an actuator add to actuator string
+        // else add to sensor string
+
+        if (isAct.get(zone)) {
+            actuators.add("{\"pin\":" + Integer.toString(pin) + "}");
+            StringType state = new StringType("0");
+            updateState(channel, state);
+        }
+
+        else {
+            sensors.add("{\"pin\":" + Integer.toString(pin) + "}");
+        }
+
+        logger.debug(sensors.toString());
+        // launch the update so the new channel settings can be sent
+        updateKonnectedModule();
+    }
+
+    @Override
+    public synchronized void channelUnlinked(ChannelUID channel) {
+        logger.debug("Channel {} has been unlinked", channel.toString());
+        // get the zone number in integer form
+        setisAct();
+        Integer zone = Integer.parseInt(channel.getId().substring((channel.getId().length() - 1)));
+        // convert the zone to the pin based on value at index of zone
+        Integer pin = Arrays.asList(PIN_TO_ZONE).get(zone);
+
+        actuators.remove("{\"pin\":" + Integer.toString(pin) + "}");
+
+        sensors.remove("{\"pin\":" + Integer.toString(pin) + "}");
+
+        logger.debug(sensors.toString());
+        updateKonnectedModule();
+    }
+
+    /*
+     * This method constructs the payload that will be sent
+     * to the konnected module via the put request
+     * it adds the appropriate senors and actuators
+     * as well as the location of the callback servelet
+     * and auth_token which can be used for validation
+     *
+     */
+    private String contructSettingsPayload() {
+        String hostPath = "";
+        try {
+            hostPath = getHostName() + webHookServlet.getPath();
+            logger.debug("The host path is: {}", hostPath);
+        }
+
+        catch (UnknownHostException e) {
+            logger.debug("Unable to obtain hostname: {}", e);
+            return "none";
+        }
+
+        String authToken = config.Auth_Token;
+        logger.debug("The Auth_Token is: {}", authToken);
+        logger.debug("The Sensor String is: {}", sensors.toString());
+        logger.debug("The Actuator String is: {}", actuators.toString());
+
+        updatepinstate();
+        String payload = "{\"sensors\":" + sensors.toString() + ",\"actuators\": " + actuators.toString()
+                + ",\"token\": \"" + authToken + "\",\"apiUrl\": \"http://" + hostPath + "\"}";
```

Comment:
seems weird that you have to construct JSON manually. Is there not a cleaner/better way?

Reply:
The code would be easier to read/maintain if Gson is used for (de)serialization JSON.
You can have a look at the Nest binding how that is done.

Then you can use these data objects and no longer have to store JSON in: 

```java
private List<String> sensors;
private List<String> actuators;
```

Hopefully that also removes all the JSON from methods such as channelLinked/channelUnlinked

Reply:
so the reason for the json in the channel linked and unlinked is to track which channels a user as linked to items as a means to determine which zones should be sent to the konnected module.  The module only sends out reports via http for pins that the user has set via a json put request.  So what i did was use the channel linked and unlinke methods to generate the json payloads with the relevant settings and then transmit them once the system was done linking all channels.

If there was a way for me to directly access what channels a user had linked to what items i could bypass the need to do this in this iterative fashion.  But i was unable to see how it could be done via the framework as the bindings are blind with what items are linked to the channels.

I don't need to know anything about the item that is linked just that an item is linked to that channel and then that channel could be added as a zone.

Reply:
You could use something like the code below instead in `contructSettingsPayload()`: 

```java
getThing().getChannels().forEach(channel -> {
    if (isLinked(channel.getUID())) {
    } else {
    }
});
```

Reply:
used this code and i was able to get rid of the chanel unlinked methods completely